alfred-telemetry-platform/src/main/java/eu/xenit/alfred/telemetry/binder/solr/sharding/SolrShardingMetricsContainer.java

```diff
@@ -0,0 +1,64 @@
+package eu.xenit.alfred.telemetry.binder.solr.sharding;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.alfresco.repo.index.shard.Floc;
+import org.alfresco.repo.index.shard.Shard;
+import org.alfresco.repo.index.shard.ShardInstance;
+import org.alfresco.repo.index.shard.ShardRegistry;
+import org.alfresco.repo.index.shard.ShardRegistryImpl;
+import org.alfresco.repo.index.shard.ShardRegistryImpl.ReplicaState;
+import org.alfresco.repo.index.shard.ShardState;
+
+public class SolrShardingMetricsContainer {
+
+    private ShardRegistry shardRegistry;
+    private HashMap<Floc, HashMap<Shard, HashSet<ShardState>>> rawData;
+    private long lastRefresh;
+    private long ttl; // how long before the raw data will be refreshed
+
+    public SolrShardingMetricsContainer(ShardRegistry shardRegistry) {
+        this(shardRegistry, 60);
+    }
+
+    public SolrShardingMetricsContainer(ShardRegistry shardRegistry, long ttl) {
+        this.shardRegistry = shardRegistry;
+        this.ttl = ttl;
+        this.lastRefresh = 0;
+    }
+
+    public void refresh() {
+        long now = System.currentTimeMillis();
+        if (rawData == null || now - lastRefresh > ttl) {
+            rawData = shardRegistry.getFlocs();
+        }
+    }
+
+    public Set<Floc> getFlocs() {
+        return rawData.keySet();
+    }
+
+    public Set<Shard> getShards(Floc floc) {
+        return rawData.get(floc).keySet();
+    }
+
+    public Set<ShardInstance> getShardInstances(Shard shard) {
+        return rawData.get(shard.getFloc()).get(shard).stream().map(shardState -> shardState.getShardInstance())
+                .collect(
+                        Collectors.toSet());
+    }
+
+    public ShardState getShardState(ShardInstance shardInstance) {
```

Comment:
See https://github.com/xenit-eu/alfred-telemetry/issues/36
Should this be converted this to `Optional<ShardState>` ?